### PR TITLE
Continue executing the script if queuing repo migrations fails

### LINKED ISSUE #781: 502 Bad Gateway can cause the migration to halt
## Description

We attempted to run a trial migration over the weekend via the CLI (v0.31); it got ~4 hours into queuing repos and failed with:

```
[9:16 pm] [INFO] You are running the latest version of the gei CLI [v0.31]
[9:16 pm] [INFO] Migrating Repo...
[9:16 pm] [INFO] GITHUB SOURCE ORG:<ourorg>
[9:16 pm] [INFO] SOURCE REPO: <repo>
[9:16 pm] [INFO] GITHUB TARGET ORG: <ourorg>-target-trial-2
[9:16 pm] [INFO] TARGET REPO: <repo>
[9:16 pm] [ERROR] An unexpected error happened. Please see the logs for details.
```

The verbose logs show:

```
9:16 pm] [DEBUG] HTTP GET: https://raw.githubusercontent.com/github/gh-gei/main/LATEST-VERSION.txt
[9:16 pm] [DEBUG] RESPONSE (OK): v0.31

[9:16 pm] [INFO] You are running the latest version of the gei CLI [v0.31]
[9:16 pm] [INFO] Migrating Repo...
[9:16 pm] [INFO] GITHUB SOURCE ORG: <ourorg>
[9:16 pm] [INFO] SOURCE REPO: <repo>
[9:16 pm] [INFO] GITHUB TARGET ORG: <ourorg>-target-trial-2
[9:16 pm] [INFO] TARGET REPO: <repo>
[9:16 pm] [DEBUG] HTTP POST: https://api.github.com/graphql
[9:16 pm] [DEBUG] HTTP BODY: {"query":"query($login: String!) {organization(login: $login) { login, id, name } }","variables":{"login":"<ourorg>-target-trial-2"}}
[9:16 pm] [DEBUG] GITHUB REQUEST ID: F87D:52D2:159A400:16FBA90:63C12F78
[9:16 pm] [DEBUG] RESPONSE (OK): {"data":{"organization":{"login":"<ourorg>-target-trial-2","id":"O_kgDOB06WWA","name":"<ourorg>-target-trial-2"}}}
[9:16 pm] [DEBUG] HTTP POST: https://api.github.com/graphql
[9:16 pm] [DEBUG] HTTP BODY: {"query":"mutation createMigrationSource($name: String!, $url: String!, $ownerId: ID!, $type: MigrationSourceType!) { createMigrationSource(input: {name: $name, url: $url, ownerId: $ownerId, type: $type}) { migrationSource { id, name, url, type } } }","variables":{"name":"GHEC Source","url":"https://github.com","ownerId":"O_kgDOB06WWA","type":"GITHUB_ARCHIVE"},"operationName":"createMigrationSource"}
[9:16 pm] [DEBUG] GITHUB REQUEST ID: F87D:52D2:159A42B:16FBAB0:63C12F78
[9:16 pm] [DEBUG] RESPONSE (OK): {"data":{"createMigrationSource":{"migrationSource":{"id":"MS_kgDaACRmMDBiZThhNS0zMzIyLTRkYzUtODliMS03Zjg1ZjExY2E2YWU","name":"GHEC Source","url":"https://github.com","type":"GITHUB_ARCHIVE"}}}}
[9:16 pm] [DEBUG] HTTP POST: https://api.github.com/graphql
[9:16 pm] [DEBUG] HTTP BODY: {"query":"\n                mutation startRepositoryMigration(\n                    $sourceId: ID!,\n                    $ownerId: ID!,\n                    $sourceRepositoryUrl: URI!,\n                    $repositoryName: String!,\n                    $continueOnError: Boolean!,\n                    $gitArchiveUrl: String,\n                    $metadataArchiveUrl: String,\n                    $accessToken: String!,\n                    $githubPat: String,\n                    $skipReleases: Boolean,\n                    $lockSource: Boolean) { \n                startRepositoryMigration(\n                    input: { \n                        sourceId: $sourceId,\n                        ownerId: $ownerId,\n                        sourceRepositoryUrl: $sourceRepositoryUrl,\n                        repositoryName: $repositoryName,\n                        continueOnError: $continueOnError,\n                        gitArchiveUrl: $gitArchiveUrl,\n                        metadataArchiveUrl: $metadataArchiveUrl,\n                        accessToken: $accessToken,\n                        githubPat: $githubPat,\n                        skipReleases: $skipReleases,\n                        lockSource: $lockSource\n                    }\n                ) {\n                    repositoryMigration {\n                        id,\n                        migrationSource {\n                            id,\n                            name,\n                            type\n                        },\n                        sourceUrl,\n                        state,\n                        failureReason\n                    }\n                  } }","variables":{"sourceId":"MS_kgDaACRmMDBiZThhNS0zMzIyLTRkYzUtODliMS03Zjg1ZjExY2E2YWU","ownerId":"O_kgDOB06WWA","sourceRepositoryUrl":"https://github.com/<ourorg>/<repo>","repositoryName":"<repo>","continueOnError":true,"gitArchiveUrl":null,"metadataArchiveUrl":null,"accessToken":"***","githubPat":"***","skipReleases":false,"lockSource":false},"operationName":"startRepositoryMigration"}
[9:16 pm] [DEBUG] GITHUB REQUEST ID: F87D:52D2:159A44C:16FBADB:63C12F79
[9:16 pm] [DEBUG] RESPONSE (BadGateway): {
   "data": null,
   "errors":[
      {
         "message":"Something went wrong while executing your query. This may be the result of a timeout, or it could be a GitHub bug. Please include `F87D:52D2:159A44C:16FBADB:63C12F79` when reporting this issue."
      }
   ]
}

[9:16 pm] [ERROR] [HTTP ERROR 502] System.Net.Http.HttpRequestException: Response status code does not indicate success: 502 (Bad Gateway).
   at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode()
   at OctoshiftCLI.GithubClient.SendAsync(HttpMethod httpMethod, String url, Object body, HttpStatusCode expectedStatus, Dictionary`2 customHeaders)
   at OctoshiftCLI.GithubClient.PostAsync(String url, Object body, Dictionary`2 customHeaders)
   at OctoshiftCLI.GithubApi.StartMigration(String migrationSourceId, String sourceRepoUrl, String orgId, String repo, String sourceToken, String targetToken, String gitArchiveUrl, String metadataArchiveUrl, Boolean skipReleases, Boolean lockSource)
   at OctoshiftCLI.GithubEnterpriseImporter.Handlers.MigrateRepoCommandHandler.Handle(MigrateRepoCommandArgs args)
   at OctoshiftCLI.Extensions.CommandExtensions.<>c__DisplayClass1_0`3.<<ConfigureCommand>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Invocation.AnonymousCommandHandler.InvokeAsync(InvocationContext )
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass17_0.<<UseParseErrorReporting>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass12_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseVersionOption>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass19_0.<<UseTypoCorrections>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__18_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseParseDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__5_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass8_0.<<UseExceptionHandler>b__0>d.MoveNext()

```

As a result, not all repos were queued for migration.

### LINKED ISSUE #840: Make generated script resilient to errors
Previously we did some work to make the generated script more robust, including a script to run migrations in parallel. In the parallel script version we kickoff all migrations and asynchronously wait for them to finish with `wait-for-migration` command. In this approach a migration failure shouldn't halt the entire execution of the script and the script already does so except where it couldn't execute a migration kick-off command! This causes some unwanted behavior and halts the entire script execution. We need to basically not exit the execution when executing the migrations.  

## Todo
- [x] Update `gei generate-script` to not exit if queueing fails for a migration.
- [x] Update `ado2gh generate-script` to not exit if queueing fails for a migration.

### PR DESCRIPTION
**Pull request by ArinGhazarian**
Closes #840 
Fixes #781 

- [x] Did you write/update appropriate tests
- [x] Release notes updated (if appropriate)
- [x] Appropriate logging output
- [x] Issue linked
- [x] Docs updated (or issue created)
- [x] New package licenses are added to `ThirdPartyNotices.txt` (if applicable)

### Description

This PR makes a generated script by `ado2gh` and `gei` for parallel migrations more resilient by not halting the entire script execution if a repo migration fails in the queueing phase. This change only applies to scripts generated to perform migrations in parallel. Also I tired to keep the changes as minimal and simple as possible since one of our main goals is to keep the script easy to understand and modify by users. 

### Before/After Script Samples

<details><summary>ADO2GH Before</summary>
<p>

```powershell
#!/usr/bin/env pwsh

# =========== Created with CLI version 1.0.0.0 ===========

function Exec {
    param (
        [scriptblock]$ScriptBlock
    )
    & @ScriptBlock
    if ($lastexitcode -ne 0) {
        exit $lastexitcode
    }
}

function ExecAndGetMigrationID {
    param (
        [scriptblock]$ScriptBlock
    )
    $MigrationID = Exec $ScriptBlock | ForEach-Object {
        Write-Host $_
        $_
    } | Select-String -Pattern "\(ID: (.+)\)" | ForEach-Object { $_.matches.groups[1] }
    return $MigrationID
}

function ExecBatch {
    param (
        [scriptblock[]]$ScriptBlocks
    )
    $Global:LastBatchFailures = 0
    foreach ($ScriptBlock in $ScriptBlocks)
    {
        & @ScriptBlock
        if ($lastexitcode -ne 0) {
            $Global:LastBatchFailures++
        }
    }
}

$Succeeded = 0
$Failed = 0
$RepoMigrations = [ordered]@{}

# =========== Queueing migration for Organization: adoorg ===========

# === Queueing repo migrations for Team Project: adoorg/BarProject ===

$MigrationID = ExecAndGetMigrationID { gh ado2gh migrate-repo --ado-org "adoorg" --ado-team-project "BarProject" --ado-repo "BarProject" --github-org "GitHubOrg" --github-repo "BarProject-BarProject" }
$RepoMigrations["adoorg/BarProject-BarProject"] = $MigrationID

$MigrationID = ExecAndGetMigrationID { gh ado2gh migrate-repo --ado-org "adoorg" --ado-team-project "BarProject" --ado-repo "DragonRepo" --github-org "GitHubOrg" --github-repo "BarProject-DragonRepo" }
$RepoMigrations["adoorg/BarProject-DragonRepo"] = $MigrationID

# =========== Waiting for all migrations to finish for Organization: adoorg ===========

# === Waiting for repo migration to finish for Team Project: BarProject and Repo: BarProject. Will then complete the below post migration steps. ===
$CanExecuteBatch = $true
if ($null -ne $RepoMigrations["adoorg/BarProject-BarProject"]) {
    gh ado2gh wait-for-migration --migration-id $RepoMigrations["adoorg/BarProject-BarProject"]
    $CanExecuteBatch = ($lastexitcode -eq 0)
}
if ($CanExecuteBatch) {
    $Succeeded++
} else {
    $Failed++
}

# === Waiting for repo migration to finish for Team Project: BarProject and Repo: DragonRepo. Will then complete the below post migration steps. ===
$CanExecuteBatch = $true
if ($null -ne $RepoMigrations["adoorg/BarProject-DragonRepo"]) {
    gh ado2gh wait-for-migration --migration-id $RepoMigrations["adoorg/BarProject-DragonRepo"]
    $CanExecuteBatch = ($lastexitcode -eq 0)
}
if ($CanExecuteBatch) {
    $Succeeded++
} else {
    $Failed++
}

Write-Host =============== Summary ===============
Write-Host Total number of successful migrations: $Succeeded
Write-Host Total number of failed migrations: $Failed

if ($Failed -ne 0) {
    exit 1
}
```

</p>
</details>

<details><summary>ADO2GH After</summary>
<p>

```powershell
#!/usr/bin/env pwsh

# =========== Created with CLI version 1.0.0.0 ===========

function Exec {
    param (
        [scriptblock]$ScriptBlock
    )
    & @ScriptBlock
    if ($lastexitcode -ne 0) {
        exit $lastexitcode
    }
}

function ExecAndGetMigrationID {
    param (
        [scriptblock]$ScriptBlock
    )
    $MigrationID = & @ScriptBlock | ForEach-Object {
        Write-Host $_
        $_
    } | Select-String -Pattern "\(ID: (.+)\)" | ForEach-Object { $_.matches.groups[1] }
    return $MigrationID
}

function ExecBatch {
    param (
        [scriptblock[]]$ScriptBlocks
    )
    $Global:LastBatchFailures = 0
    foreach ($ScriptBlock in $ScriptBlocks)
    {
        & @ScriptBlock
        if ($lastexitcode -ne 0) {
            $Global:LastBatchFailures++
        }
    }
}

$Succeeded = 0
$Failed = 0
$RepoMigrations = [ordered]@{}

# =========== Queueing migration for Organization: adoorg ===========

# === Queueing repo migrations for Team Project: adoorg/BarProject ===

$MigrationID = ExecAndGetMigrationID { gh ado2gh migrate-repo --ado-org "adoorg" --ado-team-project "BarProject" --ado-repo "BarProject" --github-org "GitHubOrg" --github-repo "BarProject-BarProject" }
$RepoMigrations["adoorg/BarProject-BarProject"] = $MigrationID

$MigrationID = ExecAndGetMigrationID { gh ado2gh migrate-repo --ado-org "adoorg" --ado-team-project "BarProject" --ado-repo "DragonRepo" --github-org "GitHubOrg" --github-repo "BarProject-DragonRepo" }
$RepoMigrations["adoorg/BarProject-DragonRepo"] = $MigrationID

# =========== Waiting for all migrations to finish for Organization: adoorg ===========

# === Waiting for repo migration to finish for Team Project: BarProject and Repo: BarProject. Will then complete the below post migration steps. ===
$CanExecuteBatch = $false
if ($null -ne $RepoMigrations["adoorg/BarProject-BarProject"]) {
    gh ado2gh wait-for-migration --migration-id $RepoMigrations["adoorg/BarProject-BarProject"]
    $CanExecuteBatch = ($lastexitcode -eq 0)
}
if ($CanExecuteBatch) {
    $Succeeded++
} else {
    $Failed++
}

# === Waiting for repo migration to finish for Team Project: BarProject and Repo: DragonRepo. Will then complete the below post migration steps. ===
$CanExecuteBatch = $false
if ($null -ne $RepoMigrations["adoorg/BarProject-DragonRepo"]) {
    gh ado2gh wait-for-migration --migration-id $RepoMigrations["adoorg/BarProject-DragonRepo"]
    $CanExecuteBatch = ($lastexitcode -eq 0)
}
if ($CanExecuteBatch) {
    $Succeeded++
} else {
    $Failed++
}

Write-Host =============== Summary ===============
Write-Host Total number of successful migrations: $Succeeded
Write-Host Total number of failed migrations: $Failed

if ($Failed -ne 0) {
    exit 1
}
```

</p>
</details>

----

<details><summary>GEI Before</summary>
<p>

```powershell
#!/usr/bin/env pwsh

# =========== Created with CLI version 1.0.0.0 ===========

function Exec {
    param (
        [scriptblock]$ScriptBlock
    )
    & @ScriptBlock
    if ($lastexitcode -ne 0) {
        exit $lastexitcode
    }
}

function ExecAndGetMigrationID {
    param (
        [scriptblock]$ScriptBlock
    )
    $MigrationID = Exec $ScriptBlock | ForEach-Object {
        Write-Host $_
        $_
    } | Select-String -Pattern "\(ID: (.+)\)" | ForEach-Object { $_.matches.groups[1] }
    return $MigrationID
}

$Succeeded = 0
$Failed = 0
$RepoMigrations = [ordered]@{}

# =========== Organization: github-source-org ===========

# === Queuing repo migrations ===
$MigrationID = ExecAndGetMigrationID { gh gei migrate-repo --github-source-org "github-source-org" --source-repo "repo-1" --github-target-org "github-target-org" --target-repo "repo-1" --ghes-api-url "https://myghes.com/api/v3" }
$RepoMigrations["repo-1"] = $MigrationID

$MigrationID = ExecAndGetMigrationID { gh gei migrate-repo --github-source-org "github-source-org" --source-repo "repo-2" --github-target-org "github-target-org" --target-repo "repo-2" --ghes-api-url "https://myghes.com/api/v3" }
$RepoMigrations["repo-2"] = $MigrationID


# =========== Waiting for all migrations to finish for Organization: github-source-org ===========

gh gei wait-for-migration --migration-id $RepoMigrations["repo-1"]
if ($lastexitcode -eq 0) { $Succeeded++ } else { $Failed++ }

gh gei wait-for-migration --migration-id $RepoMigrations["repo-2"]
if ($lastexitcode -eq 0) { $Succeeded++ } else { $Failed++ }


Write-Host =============== Summary ===============
Write-Host Total number of successful migrations: $Succeeded
Write-Host Total number of failed migrations: $Failed

if ($Failed -ne 0) {
    exit 1
}
```

</p>
</details>

<details><summary>GEI After</summary>
<p>

#### yes, even hidden code blocks!

```powershell
#!/usr/bin/env pwsh

# =========== Created with CLI version 1.0.0.0 ===========

function ExecAndGetMigrationID {
    param (
        [scriptblock]$ScriptBlock
    )
    $MigrationID = & @ScriptBlock | ForEach-Object {
        Write-Host $_
        $_
    } | Select-String -Pattern "\(ID: (.+)\)" | ForEach-Object { $_.matches.groups[1] }
    return $MigrationID
}

$Succeeded = 0
$Failed = 0
$RepoMigrations = [ordered]@{}

# =========== Organization: github-source-org ===========

# === Queuing repo migrations ===
$MigrationID = ExecAndGetMigrationID { gh gei migrate-repo --github-source-org "github-source-org" --source-repo "repo-1" --github-target-org "github-target-org" --target-repo "repo-1" --ghes-api-url "https://myghes.com/api/v3" }
$RepoMigrations["repo-1"] = $MigrationID

$MigrationID = ExecAndGetMigrationID { gh gei migrate-repo --github-source-org "github-source-org" --source-repo "repo-2" --github-target-org "github-target-org" --target-repo "repo-2" --ghes-api-url "https://myghes.com/api/v3" }
$RepoMigrations["repo-2"] = $MigrationID


# =========== Waiting for all migrations to finish for Organization: github-source-org ===========

if ($RepoMigrations["repo-1"]) { gh gei wait-for-migration --migration-id $RepoMigrations["repo-1"] }
if ($RepoMigrations["repo-1"] -and $lastexitcode -eq 0) { $Succeeded++ } else { $Failed++ }

if ($RepoMigrations["repo-2"]) { gh gei wait-for-migration --migration-id $RepoMigrations["repo-2"] }
if ($RepoMigrations["repo-2"] -and $lastexitcode -eq 0) { $Succeeded++ } else { $Failed++ }


Write-Host =============== Summary ===============
Write-Host Total number of successful migrations: $Succeeded
Write-Host Total number of failed migrations: $Failed

if ($Failed -ne 0) {
    exit 1
}
```

</p>
</details>

### CODE REVIEW

File: RELEASENOTES.md [hunk 1]
```diff
@@ -1 +1,2 @@
-- Create shared access signature (SAS) with read-only permissions - not read-write - when generating Azure Blob Storage URL
\ No newline at end of file
+- Create shared access signature (SAS) with read-only permissions - not read-write - when generating Azure Blob Storage URL
+- Make generated script by `gh gei generate-script` and `gh ado2gh generate-script` commands for parallel migrations more resilient by not halting the entire script executing if queuing a repo migration fails.
```
Comment (timrogers) [inline RELEASENOTES.md:2]:
```suggestion
- Make parallel migrations scripts generated by `gh gei generate-script` and `gh ado2gh generate-script` more resilient by not halting the entire script if queuing a repo migration fails. 
```

Comment (timrogers) [inline RELEASENOTES.md:2]: What would it look like to apply a similar change to `bbs2gh`?

**Review by timrogers — COMMENTED** [review]
I won't try to review the code here - happy to trust @dylan-smith's 👀 much more than I'd trust my own!

File: RELEASENOTES.md (hunk 1 above)
Comment (ArinGhazarian) [reply RELEASENOTES.md:2]: `bbs2gh` doesn't have a parallel execution yet!

**Review by dylan-smith — APPROVED** [review]

**Pull request merged**